Add centered pressure gradient implimentation

sbrus89 opened the pull request.

xylar commented on components/omega/src/ocn/Eos.cpp:
```diff
@@ -140,6 +144,72 @@ void Eos::computeSpecVol(const Array2DReal &ConservTemp,
+   parallelFor(
+       "compute-interface-values", {NCellsAll, NVertLayers + 1},
+       KOKKOS_LAMBDA(I4 ICell, I4 KInterface) {
+          if (KInterface == 0) {
```
As we discussed, we need hierarchical parallelism here and this should be `MinLayerCell` or whatever the proper variable is, rather than 0.

sbrus89 replied:
Yes, I'll take care of swapping out for hierachical parallelism.

sbrus89 committed:
Udpate pressure gradient

 - Average specific volume to interfaces
 - Average product of spcific volume and pressure to edge
 - Add centered and high order functor constructors
 - Format linting

sbrus89 committed:
Add test driver

overfelt committed:
Cleanup branch starting with new commit.

overfelt committed:
Rename flag used to trigger higher order advection.

overfelt committed:
Rename flag used to trigger higher order advection.

overfelt committed:
Rename flag used to trigger higher order advection.

overfelt committed:
Rename flag used to trigger higher order advection.

overfelt committed:
Clang format.

overfelt committed:
Clang format with newer version.

sbrus89 committed:
Add PressureGrad section to Default.yml

sbrus89 committed:
Updates to PGradTest

sbrus89 committed:
Add tilt factor to PGradTest

xylar commented on components/omega/src/ocn/Eos.cpp:
```diff
@@ -140,6 +144,79 @@ void Eos::computeSpecVol(const Array2DReal &ConservTemp,
+             // AbsSalinityInterface(ICell, KInterface) =
+             //     AbsSalinity(ICell, KInterface);
+             SpecVolInterface(ICell, KInterface) = SpecVol(ICell, KInterface);
+          } else if (KInterface == NVertLayers + 1) {
```
Similarly, this should have hierarchical parallelism and use `MaxLayerCell` or similar.

xylar commented on components/omega/src/ocn/Eos.cpp:
```diff
@@ -140,6 +144,79 @@ void Eos::computeSpecVol(const Array2DReal &ConservTemp,
+/// Compute specific volume for all layer interfaces(no displacement)
+void Eos::computeSpecVolInterface(const Array2DReal &ConservTemp,
+                                  const Array2DReal &AbsSalinity,
+                                  const Array2DReal &PressureInterface) {
```
We have to decide if we want this.  My feeling would be that we wouldn't want the extra cost of calling the EOS at each layer interface and that we would instead average the cell-center, layer-mid-point values to the layer interfaces.

sbrus89 replied:
Agreed, I will eliminate this.

xylar commented on components/omega/src/ocn/PGrad.h:
```diff
@@ -0,0 +1,178 @@
+      const I4 ICell1      = CellsOnEdge(IEdge, 1);
+      const Real InvDcEdge = 1.0_Real / DcEdge(IEdge);
+
+      for (int KVec = 0; KVec < VecLength; ++KVec) {
```
This needs to be over something like `MinLayerEdgeBot` to `MaxLayerEdgeTop` with hierarchical parallelism.

xylar commented on components/omega/src/ocn/PGrad.h:
```diff
@@ -0,0 +1,178 @@
+         const Real PAlphaEdgeK =
+             0.5_Real *
+             (PressureInterface(ICell1, K) * SpecVolInterface(ICell1, K) +
+              PressureInterface(ICell0, K) * SpecVolInterface(ICell0, K));
```
Let's chat about it but I think we want the `PAlpha` product to be averaged both to cell centers and to layer interfaces.

sbrus89 replied:
Check out my latest commit and let me know what you think. The new averaging appeared to lower errors in my hydrostatic test.

xylar reviewed (commented):
@sbrus89, a few comments. The main ones in terms of needing to discuss them relate to how the product of $\alpha p$ should be computed  at layer interfaces and edges.  I think it probably needs to be computed based on 4-point averages from cell centers and layer midpoints but we should make sure we're on the same page.

sbrus89 committed:
Test for pressure gradient convergence in PGradTest

sbrus89 committed:
Update interface product averaging

sbrus89 committed:
Increase number of refinements in convergence test loop

sbrus89 committed:
Remove debug comments

sbrus89 committed:
Run PGrad test in serial

sbrus89 committed:
Add centered pressure gradient implimentation

sbrus89 committed:
Convert Pa to dbar in Teos10Eos functor

sbrus89 committed:
Remove SpecVolInterface calculation

overfelt committed:
Add documentation for the higher order horizontal flux calculation.

overfelt committed:
Remove whitespace the linter disapproves of.

sbrus89 committed:
Fix error calculation to exclude top and bottom levels

sbrus89 committed:
Convert pressure to Pa in EosTest

sbrus89 committed:
Switch to hierarchical parallelism

xylar commented:
@sbrus89, your last few commits look reasonable to me.  I would remove the commented-out loops after the switch to hierarchical parallelism eventually but it's okay to keep them for now while you make sure things are working as expected.

It's getting cumbersome to review here because your local omega/develop is out of date, so let's just got ahead and make the PR to Omega.

sbrus89 committed:
Fix bounds for bottom interface product

sbrus89 committed:
Remove debug comments and cleanup

sbrus89 committed:
Clean up

sbrus89 committed:
Fix for gradient in interface term

overfelt committed:
Fix off-by-one array size in unit test.

overfelt committed:
Fix a couple of signed/unsigned int comparison compiler warnings.

overfelt committed:
Fix Kokkos View host/device incompatibility errors.

overfelt committed:
Fix up use of VCoord after rebase from develop.

overfelt committed:
Clang format

xylar committed:
Fix incorrect $\rho_0$ multipliers up to (37)

Surface flux terms, body forces and pressure terms were all getting
multiplied by $\rho_0$, leading to incorrect dimensions and
inconsistency with the rest of the derivation.

xylar committed:
More fixes to the momentum derivation

xylar committed:
Clarify the horizontal gradient of Phi

xylar committed:
Fix more in Layer Equations and Discrete Equations

xylar committed:
Rewrite PGF throughout sections 9 and 10

sbrus89 committed:
Add Liebnez terms for geopotential gradient

sbrus89 committed:
Montgomery potential implementation

sbrus89 committed:
Work toward outputing PGrad tendencies (need to fix unit tests)

brian-oneill committed:
VertAdv module definition; creation, destruction, retrieval methods

brian-oneill committed:
Add optional per-team scratch space to parallelForOuter

brian-oneill committed:
Add computeVerticalVelocity

brian-oneill committed:
Add computeThicknessVAdvTend

brian-oneill committed:
Add computeVelocityVAdvTend

brian-oneill committed:
Add computeTracerVAdvTend dispatch function

brian-oneill committed:
Add computeVerticalFluxes

brian-oneill committed:
Add computeStdVAdvTend

brian-oneill committed:
Add computeFCTVAdvTend

overfelt committed:
Clang Format.

overfelt committed:
Fix compiler errors after merge from develop.

sbrus89 committed:
Fix Driver and Timestepper ctests

brian-oneill committed:
Add ProvThickness to auxvars (follow-up restructuring needed)

sbrus89 committed:
Fix Tendencies ctest

sbrus89 committed:
Fix some rebase issues

 - EdgeMask moved to VertCoord
 = Density0 no longer in omega.yaml

sbrus89 committed:
Compute BruntVaisalaFreqSq with input pressure in Pa

sbrus89 committed:
Fix EosTest, pressure dbar -> Pa conversion

sbrus89 committed:
Fix Tendencies ctest

brian-oneill committed:
Add new config variables to default config

brian-oneill committed:
Add VertAdv unit tests

brian-oneill committed:
Init some previously missed pointers and add Field for ProvThickness

brian-oneill committed:
Comments, bug fix, and add Fields for VertAdv arrays

brian-oneill committed:
Some bug fixes for running on GPU

brian-oneill committed:
Bug fixes: BottomDepth no longer in HorzMesh and typo fixes

brian-oneill committed:
Make host copies of VertAdv arrays

brian-oneill committed:
Linting fix

overfelt committed:
Forgot to zero out tendecies before summing into values.

overfelt committed:
Fix for parallel: NCellsOwned -> NCellsAll/

brian-oneill committed:
Resize flux arrays in VertAdvTest

brian-oneill committed:
Rename config option from Enabled to Enable

brian-oneill committed:
Remove unnecessary header

brian-oneill committed:
Skip vertical tracer advection when not enabled

xylar commented on components/omega/src/ocn/Tendencies.cpp:
We also need to *call* `defineFields()`:
```suggestion
    NTracers = NTracersIn;

   defineFields();

```
(Note: I think GitHub suggestion is glitching on indentation lately.  Better to make the change manually than commit the suggestion.)

sbrus89 replied:
Good catch, thanks

mwarusz committed:
Exchange tracer halos at every RK stage

xylar commented on components/omega/test/ocn/PGradTest.cpp:
```diff
@@ -0,0 +1,310 @@
+
+            // compute specific volume from EOS
+            VCoord->computePressure(LayerThick, SurfacePressure);
+            DefEos->computeSpecVol(Temp, Salinity, PressureMid);
```
This is probably not where we want to be computing the specific volume in the long run, right?  We would want to compute it somewhere external in the time step and just make it available here?  I guess that's what Hyun is working on?

sbrus89 replied:
Right, I realized I need to add this call in Tendencies as well.

mwarusz committed:
Fix conservation issue in high order advection

sbrus89 committed:
Add computation of SpecVol to Tendencies

 - Requires compptation of pressure

sbrus89 committed:
Read PressureGradTendencyEnable and set PGrad enable flag

brian-oneill committed:
Add TimeStep to AuxState for ProvThickness

mwarusz committed:
Use accessor functions to get state variables

brian-oneill committed:
Remove commented out ssh calculation

brian-oneill committed:
Add LayerThicknessAux.computeVarsOnCells to AuxState::computeAll

This PR adds ProvThickness to LayerThicknessAux, which is
needed by the tracer equations. PR #335 removes SSH from
LayerThicknessAux and removes computeVarsOnCells from
computeMomAux.

brian-oneill committed:
Add VertAdv to Tendencies

brian-oneill committed:
Disable VertAdv tends in TimeStepperTest

brian-oneill committed:
Add VertAdv to AuxState

brian-oneill committed:
Add computeVerticalVelocity to computeMomAux

brian-oneill committed:
Add VertAdv init and clear to OceanDriver methods

brian-oneill committed:
Add computeThicknessVAdvTend to computeThicknessTendenciesOnly

brian-oneill committed:
Add computeVelocityVAdvTend to computeVelocityTendenciesOnly

brian-oneill committed:
Add computeTracerVAdvTend to computeTracerTendenciesOnly

Also add TimeStep to Tendencies

brian-oneill committed:
Make TimeStep nonoptional for computeTracerVAdvTend

brian-oneill committed:
Add VertAdv devGuide and userGuide

mwarusz committed:
Require valid time index in state getters

mwarusz committed:
Refactor Tracer::getByIndex and Tracers::getByName

mwarusz committed:
Require valid time index in tracer getters

Copilot committed:
Update OceanState and Tracers documentation to reflect new getter API

brian-oneill committed:
Add early return to VertAdv methods when NVertLayers == 1

xylar committed:
Lintng of Tendencies.cpp

xylar committed:
Add temporary call to computeZHeight() in pgrad tendency

xylar commented on components/omega/src/ocn/PGrad.h:
```diff
@@ -0,0 +1,186 @@
+      const I4 ICell1      = CellsOnEdge(IEdge, 1);
+      const Real InvDcEdge = 1.0_Real / DcEdge(IEdge);
+      Real Rho0   = 1026.0_Real;
+      Real Gravity = 9.80616_Real;
```
This is not the right Gravity.  We should use the value from GlobalConstants.h, which comes from here:
https://github.com/E3SM-Project/Omega/blob/develop/share/util_cxx/pcd_const.h#L28

xylar commented on components/omega/src/ocn/PGrad.h:
```diff
@@ -0,0 +1,186 @@
+      const I4 ICell0      = CellsOnEdge(IEdge, 0);
+      const I4 ICell1      = CellsOnEdge(IEdge, 1);
+      const Real InvDcEdge = 1.0_Real / DcEdge(IEdge);
+      Real Rho0   = 1026.0_Real;
```
This isn't used and should be removed.

xylar committed:
Compute geometric z after specific volume

xylar committed:
Lint pressure-gradient code

xylar committed:
Use Gravity from GlobalConstants in PGrad

Drop unused Rho0

overfelt committed:
Clang format.

sbrus89 committed:
Initialize/clear Tendencies in IOStreamTest

brian-oneill committed:
Fix typos in comments and docs

brian-oneill committed:
Update HorzMesh user doc

brian-oneill committed:
Add sanity check for bottom depth

brian-oneill committed:
Fix linting fail

brian-oneill committed:
Rename TracerTend to LocTracerTend in function call

mwarusz committed:
Update Tracers dev docs

mwarusz committed:
Fix whitespace

overfelt committed:
Fix for unit test. Use actual earth radius: REarth.

overfelt committed:
Fix setting of TracerHorzAdvTendencyEnable flag based on code review.

overfelt committed:
Clang format.

sbrus89 committed:
Add gradient of tidal potential and SAL

sbrus89 committed:
Fix creation of tendency fields with same name

sbrus89 committed:
Initialize pressure gradient in state ctest

sbrus89 committed:
Fix IOStreamTest

xylar committed:
Merge pull request #326 from xylar/omega/docs/fix-omega-v1-equations

Fixes to Omega v1 Governing Equations

This pull request updates the documentation in OmegaV1GoverningEqns.md to improve the consistency, clarity, and correctness of the mathematical notation used in the derivation of the governing equations. The changes primarily remove unnecessary factors of \rho_0 from several equations, clarify the definition and notation for surface forces, and correct integration limits and variable usage for consistency.

Mathematical and Notational Corrections:

* Removed unnecessary factors of \rho_0 from the pseudo-height mass, tracer, and momentum conservation equations, as well as from their layer-averaged and Reynolds-averaged forms, to ensure dimensionally consistent and simplified expressions.
* Clarified the definition of the surface force (traction vector) as \tau = \sigma \cdot {\bf n} and distinguished the horizontal component \tau_\perp in the momentum equation, improving the physical interpretation and accuracy of boundary force terms.

Consistency and Integration Limit Corrections:

* Corrected integration limits and variable usage in the layer and Reynolds-averaged tracer equations to ensure consistency with the rest of the document and the intended physical meaning.
* Updated several equations to use consistent notation for averages, fluctuations, and layer integrals, improving readability and reducing the risk of confusion.

sbrus89 committed:
Fix PGradTest for GPU

sbrus89 committed:
Add pass/fail criteria to PGradTest

brian-oneill committed:
Remove commented-out line

sbrus89 committed:
Lint files

overfelt committed:
Change parameter name from TracerHorzAdvTendencyOrder to HorzTracerFluxOrder for consistentsy.

overfelt committed:
Changing the variable names to PascalCase to follow Omega’s naming conventions.

vanroekel committed:
Updates related to turbulence quantities

This removes confusing description related to Reynolds' averaging and
discusses instead resolved and unresolved through projection operators
this removes unecessary dependence on specific averaging.

It also adds description as to why the alpha' grad p' term is ignored

The description of the vertical viscosity sections is cleaned up

mark-petersen committed:
Merge pull request #341 from mwarusz/omega/fix-time-level

Use accessor functions to get state variables

mwarusz committed:
Add ForceLowOrder option to high-order tracer advection class

mwarusz committed:
Use high-order tracer adv class in tendency terms test

mwarusz committed:
Only use high-order tracer adv in tendencies

mwarusz committed:
Rip out old tracer advection

mwarusz committed:
Rename: TracerHighOrderHorzAdv -> TracerHorzAdv

mwarusz committed:
Move some options to the advection section

mwarusz committed:
Fix size of HighOrderFlxHorz

overfelt committed:
Merge pull request #2 from mwarusz/mwarusz/omega/high-order-adv-for-james

Changes to match MPAS-O convergence rates in tracer advection tests

sbrus89 committed:
Remove commented out code for debugging

sbrus89 committed:
Remove unsed member variables

sbrus89 committed:
Fix for updated accessor state/tracer functions following rebase

sbrus89 committed:
Add user and dev documentation

mark-petersen committed:
Merge pull request #354 from mwarusz/omega/tracer-halo-exchng

Exchange tracer halos at every RK stage

sbrus89 committed:
Fix compiler warnings from unsed variables

sbrus89 committed:
Fix linting issues

sbrus89 committed:
Merge pull request #316 from overfelt/overfelt/HigherOrderTendency_Cleanup

Standard non-monotonic higher order horizontal advection copied from mpas-ocean to Omega.

brian-oneill committed:
Rebase: remove TracerAux computeVarsOnEdge

brian-oneill committed:
Rebase: fix variable name passed to computeVelocityVAdvTend

brian-oneill committed:
Rebase: fix getLayerThickness usage

vanroekel committed:
Fixes links and updates tracer SGS terms

xylar committed:
Clarify the choice of reference density for compatibility with E3SM.

xylar committed:
Merge pull request #351 from xylar/omega/docs/fix-omega-v1-momentum-derivation

Rewrite pressure-gradient force in Omega v1 equations

This merge makes a significant change to how the pressure-gradient force term is derived in the Omega v1 equations.  In the previous version, I had tried too hard to avoid the continuous equations, instead insisting on a finite-volume representation without derivatives at the beginning of the derivation.  In particular, we need $\nabla \Phi$ to be included in the body force in our derivation.

The results of these changes were used in the implementation of the horizontal pressure-gradient force (HPGF).

brian-oneill committed:
Exchange halo for BottomDepth

sbrus89 committed:
Merge pull request #328 from brian-oneill/omega/add-vertical-advection

Add the `VertAdv` module to Omega, providing data structures and methods for vertical advection. This module includes:

- Arrays for storing vertical velocity and tracer fluxes at layer interfaces.
- A method to compute the vertical velocity from the divergence of the horizontal velocity.
- Methods for computing the tendencies of thickness, horizontal velocity, and tracers due to vertical advection. The tracer tendency methods include multiple options for order of accuracy of the flux calculation, as well as algorithmic options for standard advection or flux-corrected transport (FCT).

Additional changes in this PR:

- Extend `parallelForOuter` with an optional argument to enable the use of scratch arrays in hierarchical parallelism.
- Add the auxiliary variable `ProvThickness`, which represents thickness after horizontal thickness flux. Further work required to fully implement provisional thickness.
- Remove `BottomDepth` from `HorzMesh`, this field was previously added to `VertCoord`.

sbrus89 committed:
Fix linting issues after rebase

sbrus89 committed:
Fix vertRange -> vertRangeChunked

sbrus89 committed:
Fix converence criteria in CTest

sbrus89 committed:
Fix variable naming convention in PGradTest.cpp

sbrus89 committed:
Incorporate review feedback

sbrus89 committed:
Fix linting issues

sbrus89 committed:
Use provis tracers for RK4

sbrus89 committed:
Fix clang-format issues
